river/neighbors/base.py

```diff
@@ -1,35 +1,48 @@
 import collections
-import functools
 import operator
-from typing import Any, Callable, Tuple
+import typing
 
-from river import utils
 
-__all__ = ["NearestNeighbors", "MinkowskiNeighbors"]
+class DistanceFunc(typing.Protocol):
+    def __call__(self, a: typing.Any, b: typing.Any, **kwargs) -> float:
+        ...
 
-DistanceFunc = Callable[[Any, Any], float]
 
+class FunctionWrapper:
+    """Wrapper used to make distance function work with KNNClassifier and
+    KNNRegressor.
 
-class NearestNeighbors:
-    """A basic data structure to hold nearest neighbors.
+    The k-NN-based classifier and regressor store tuples with `(x, y)`, but only
+    `x` is used for distance calculations. This wrapper makes sure `x` is accessed
+    when calculating the distances.
 
     Parameters
     ----------
+    distance_function
+        The custom distance function to be wrapped.
+    """
```

Comment:
Style: you're supposed to add a carriage return before the `"""`

Reply:
Done!